doc/user_guide/psy_data.rst

```diff
@@ -0,0 +1,294 @@
+
+``PreEnd(this)``
+++++++++++++++++
+The method ``PreEnd`` is called after all variables before the instrumented
+region have been provided.
+
+``PostStart(this)``
++++++++++++++++++++
+This is the first call after the instrumented region. It does not take
+any parameters, but the static ``PSyDataType`` instance can be used
+to store the name and number of variables if required. This will be
+followed by calls to ``ProvideVariable``, which is described above.
+
+``PostEnd(this)``
++++++++++++++++++
+This function is the last call after an instrumented region. It indicates
+that all variables have been provided.
+ 
+An example of a library using PSyData is included in PSyclone in the
+directory ``.../lib/extract/netcdf``. This library is used to extract
+kernel input- and output-parameter and store them in a NetCDF file.
+
+.. note::
+
+    Note that only the ``PreDataStart`` call takes the module-
+    and region-name as parameters. If these names are required
+    by the profiling library in different calls, they must
+    be stored in the ``PSyData`` object so that they are
+    available when required.
+
+.. _psy_data_transformation:
+
+PSyData Transformation and PSyData-Node
+---------------------------------------
+The base transformation to create the PSyData callbacks is the 
+``PSyData`` transformation:
+
+.. autoclass:: psyclone.psyir.transformations.psy_data_trans.PSyDataTrans
+    :members:
+
+It is very similar to the profile transformation and kernel extraction
+transformation. Those transforms insert a special node into the AST,
+which at code creation time will generate the code to give access to
+the data fields. Both profile and kernel extraction
+nodes use ``PSyDataNode`` as a base class, which will create the 
+PSyData calls as described above. The difference is that those
+classes will provide different parameters to the ``PSyDataNode``,
+resulting in different code being created:
+
+.. autoclass:: psyclone.psyir.nodes.psy_data_node.PSyDataNode
+    :members:
+
+The behaviour of the ``PSyDataNode`` is controlled using the ``option``
+dictionary, as documented above. If there is no variable to be provided
+(i.e both ``pre_variable_list`` and ``post_variable_list`` are empty),
+then the ``PSyDataNode`` will only create a call to ``PreStart`` and 
+``PostEnd``. This is utilised by the profiling node to make the profiling
+API libraries (see :ref:`ProfilingAPI`) independent of the infrastructure
+library, which will contain API-specific parameters in any call to
+``ProvideVariable``. It also reduces the number of calls required before
+and after the instrumented region which can affect overall
+performance and precision of any measurements, see :ref:`profiling`
+for more details.
+
+The kernel extraction node ``ExtractNode`` uses the dependency
+module to determine which variables are input- and output-parameter,
+and provides these two lists to the ``PSyDataNode``. It also uses
+``post-var-postfix``, see :ref:`psyke`.
```

Comment:
What kind of thing is ``post-var-postfix``? I couldn't find any mention of it in the psyke link.

Reply:
Indeed wrong reference. That is described in the API above, I've changes the description.

Reply:
I'm sorry but I still don't get it - I've searched and there's no mention of post-var-postfix in the text above this line. It's possible I'm being really stupid...?

Reply:
OK, the original issue was a typo in the python doc, which prevented those variables from showing up in the manual. That's fixed, I've also expanded the documentation for these options, including a more detailed description of the netcdf extractor (in psyke.rst).